the-collab-lab/tcl-64-smart-shopping-list
# Modals

## Description

~~ **STATUS UPDATE: PR IS FINALIZED** ~~ 

Create a reusable modal component to use for multiple actions, styled to match our wireframes and design.

Each modal has a `Cancel` and a `Confirm` button. To make the modal dynamic, the component takes in a `modalBody` prop for what to display in the modal, and a `confirmationAction` prop for what action the confirm button triggers. It uses the `showModal` hook (native to the library used) to control when the modal is shown vs hidden. The modals are also responsive to work for all screen sizes.

## Related Issue

Closes #38 

## Acceptance Criteria

- [x] A modal component is added
- [x] A modal is used for `Enter list token` action (to open a list)
- [x] A modal is used for `Log out` of current list
- [x] A modal is used for `Delete Item` to remove an item from the list (this replaces the original `window.confirm` popup)

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
| ✓   | :hammer: Refactoring       |
|     | :100: Add tests            |
| ✓   | :link: Update dependencies |
|     | :scroll: Docs              |

## Updates

- Installed [Tailwind Elements React library](https://tailwind-elements.com/docs/react/) to make the [modal component](https://tailwind-elements.com/docs/react/components/modal/)
- Added trash can icon to the ListItem component to use for `Delete` button
- The old `window.confirm` popup for deleting an item has been replaced with a modal
- All three modals are fully functional, and should be fully accessible
- Slight refactoring was required in order for modals to correctly render and work. For example, `handleDelete` function now lives in the List component, instead of ListItem)
- In "Leave list" modal: Added a copy button to match the List page. Added a reminder to users to save their token for later.
- Styled modals to use the secondary `Buttons` and match input styling on rest of the app.

### After



https://github.com/the-collab-lab/tcl-64-smart-shopping-list/assets/121128718/50149ce2-cd29-420e-a4be-f5d137656d00

## Testing Steps / QA Criteria


In the app, test user flows for these scenarios:
- From the Home page, use `join existing list` button to enter an existing token in order to log in
- From the Home page, use `join existing list` button but then choose `Cancel` instead of `Confirm`
- From the List page, use `delete` icon button to remove an item from the list
- From any page, use `leave list` button to log out
- Repeat any of these using the keyboard to navigate instead of the mouse
